notebook-api/notebook/notebook.py

```diff
@@ -0,0 +1,434 @@
+# Ex: If Kubernetes names running pod status (phase) "Running", and svc is running
+# we should return "Running" here
+# TODO: Inspect exceptions to distinguish between 404 and others
+
+def read_svc_status(svc_name):
+    try:
+        svc = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except:
+        return 'Deleted'
+
+
+# TODO: support lambda expressions
+# TODO: had default key as path name, but this adds branches/if statements
+# in tight loop, dont think the space savings worth it
+
+pod_paths = [
+    ['name', 'metadata.labels.name'],
+    ['pod_name', 'metadata.name'],
+    ['svc_name', 'metadata.labels.svc_name'],
+    ['pod_status', 'status.phase'],
+    ['svc_status', 'metadata.labels.svc_name', lambda x: read_svc_status(x)],
+    ['creation_date', 'metadata.creation_timestamp',
+        lambda x: x.strftime('%D')],
+    ['token', 'metadata.labels.jupyter_token']
+]
+
+
+@app.route('/api', methods=['GET'])
+def get_notebooks():
+    user_id = request.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    # svcs = k8s.list_namespaced_service(
+    #     namespace='default',
+    #     watch=False,
+    #     label_selector='user_id=' + UNSAFE_user_id_transform(user_id),
+    #     _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS).items
+    # k8s.read_namespaced_service(svc_name, 'default')
+    pods = k8s.list_namespaced_pod(
+        namespace='default', watch=False,
+        label_selector=f"user_id={UNSAFE_user_id_transform(user_id)}").items
+
+    return marshall_json(pods, pod_paths), 200
+
+# TODO: decide if need to communicate issues to user; probably just alert devs
+
+
+@app.route('/api/<svc_name>', methods=['DELETE'])
+def delete_notebook(svc_name):
+    # TODO: Is it possible to have a falsy token value and get here?
+    if not request.headers.get("User") or not svc_name:
+        return forbidden()
+
+    escp_user_id = UNSAFE_user_id_transform(request.headers.get("User"))
+
+    try:
+        svc = k8s.read_namespaced_service(svc_name, 'default')
+
+        if svc.metadata.labels['user_id'] != escp_user_id:
+            return forbidden()
+
+        del_svc(svc_name)
+
+        pods = k8s.list_namespaced_pod(
+            namespace='default', watch=False, label_selector=f"uuid={svc.metadata.labels['uuid']}").items
+
+        if len(pods) == 0:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod not found for given uuid")
+            return '', 200
+
+        if len(pods) > 1:
+            log.error(
+                f"svc_name: {svc_name} pod_uuid: {svc.metadata.labels['uuid']}: pod_uuid is not unique")
+
+        for pod in pods:
+            if pod.metadata.labels['user_id'] != escp_user_id:
+                return forbidden()
+            del_pod(pod.metadata.name)
+
+        return '', 200
+
+    except kube.client.rest.ApiException as e:
```

Comment:
If delete fails we shouldn't return a 200. I think we need more tight `try-except`s that do the right thing in each case (or don't catch anything). It seems like for the `k8s.read_namespaced_service` we should 404 if it fails. If the delete fails we should probably 500 (meaning don't try-except at all).

Reply:
Delete can fail because the pod is already deleted (and is the most likely cause). From the user's perspective that is not an error. I agree that we could provide better granularity.